### ISSUE TITLE
implement image-proxy (fetch and mininize images)

### ISSUE THREAD

**Issue by pointhi** [searx/searx#112]
currently images are provided in the result-site, without data compression or ssl-encryption.

My Idea is that searx fetch the result images, minimize it if possible and store the images for a short time localy.

**Comment by Cqoicebordel** [comment]
It would also allow a less tracked activity on searx.

When you say minimize, is that producing thumbnails for those ?

**Comment by pointhi** [comment]
yes
- it is minimizing the images to a defined size
- caching the results for 1hour or so

my problem is:
- it is working well on my localhost with bin/searx-run
- my apache2 server with uwsgi is creating 500 Errors for most of the images, but a few are working (I don't know why)

**Comment by Cqoicebordel** [comment]
It smells like a rights issue : 
- What is your working dir for the images ?
- Does this folder is owned by you, by apache2, by uwsgi or by searx user ?
- If you are not sure, try to put a chmod 777 on the working dir and test it.

If it's not that, I would look on the HTTP queries side : 
- Does searx 'get' the images ?
- What do the servers respond (what http code) ?
- Is there a transfer between the distant server and searx ?

So the important information is to know exactly **when** the 500 error is issued : before the get, before the download, before the write on disk, before the transform, before the new write on disk or before the use by searx.

(BTW, is there a write on disk before the transform ? Because it could be an issue with the size of memory used to transform a big image into a smaller one : it must first convert it to raw, and in ram, it takes a lot of place)

**Comment by pointhi** [comment]
See: #137

https://github.com/mikeboers/Flask-Images

```
What is your working dir for the images ?
```

\tmp by default

```
Does this folder is owned by you, by apache2, by uwsgi or by searx user ?
```

chmod 777 normaly, so should no problem

```
Does searx 'get' the images ?
```

Partly, some images are processed. But I haven't found a pattern which is showing the error

```
What do the servers respond (what http code) ?
```

500 Error if the image delivery doesn't work

```
So the important information is to know exactly when the 500 error is issued
```

Thats a good question. I have searched through all logs, but doesn't found anything which is causing that errors.

**Comment by Cqoicebordel** [comment]
It's the distant server that respond 500 ? Not searx ?
If that is the case, it could be because the distant server doesn't allow curl/wget/request to grab images. It could only be an issue of using the right user agent.

I will try to go in details into your code, and test some things, but beyond what I already talked about, I have no idea...

**Comment by pointhi** [comment]
```
It's the distant server that respond 500 ? Not searx ?
```

I think not, because with `bin/searx-run` every is working fine for me

**Comment by asciimoo** [comment]
Is resizing and the whole PIL/flask-images(/gunicorn? why?) dependency chain necessary?
What about just a simple image proxy function (optionally with caching)? Then, the image parsing/manipulation can be avoided and can't be a security risk.

**Comment by pointhi** [comment]
I have implemented it in this way because it was the simplest way to do it.

If someone would implement this feature directly in searx, it would be great. But the features of flask-images and of a self written solution wouldn't differ much.

thx

**Comment by asciimoo** [comment]
Okay, i start experimenting with it. A c-based lib which depends on libjpeg and friends is a big difference imho

**Comment by Cqoicebordel** [comment]
I think resizing is not only necessary, but also essential : it's a pain to load a searx image result page with each picture weighting something like 2Mo.

And in the side of dependency, I rather have a python dependency than an OS dependant lib...

I don't think it would be too hard to build it from scratch (or at least with PIL)
http://united-coders.com/christian-harms/image-resizing-tips-every-coder-should-know/

EDIT : Wait, that's already what you did. My bad.

**Comment by Cqoicebordel** [comment]
Sorry, just saw your commits.
I don't think it's the right way to go : For that cache to be useful, it would need to have two users using the same query, or a query that return the same images, in a short time span. Because if it's the same user, the browser cache would suffice.
But, if the cache would also minimize the weight of the pictures (-> ~100ko) and cache that instead (or too), that would be very useful to the user, even if there is only one user of this searx instance.

**Comment by asciimoo** [comment]
Cache and image resize are different things. It's not too hard to add caching to the current solution and resize isn't necessary if we can grab the thumbnail pictures from the engines

**Comment by Cqoicebordel** [comment]
Agree, cache and resize are different things.
But (and I'm really sorry if I sound harsh) caching without resizing is useless : there is no use case for it. Alone, it will never be useful for the user.
It could instead be worse for the user (I didn't look and grasp the meaning of all your code so I could be wrong. If it's the case, ignore this point) :
1. The user makes an image search.
2. Searx return the url of the images from the distant server, while caching the images (it's this step I'm not sure at all)
3. The user makes the same search, and gets the cached urls of the images, breaking the browser caching.

For the "getting the thumbnails from the engines", I would gladly look at it tomorrow (that's something I would love to do). But there are some issues :
- Not all engines provide thumbnails
- Some engines provide thumbnails that are too small to be used in searx
- Some provide multiple sizes. Which ones to use ? (that the simplest case)

So, in some cases, resizing would still be good.

**Comment by dalf** [comment]
Side note : take care not to create a open proxy.

**Comment by pointhi** [comment]
- if we hash the `url` +  `secret_key`, we can add a protection to prevent misuse of the proxy. (adding a `key` property to the url)
- probably we should support subdomains (like a.searx.me, b.searx.me,...) to improve image-loading-time. because most web-browsers are loading a maximum of 2 images per domain at the same time (https://stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser). The problem is, that every subdomain require a certificate, or there is a wildcard-cert for all used searx-domains.

good work @asciimoo :+1:

**Comment by dalf** [comment]
With the SPDY protocol, domain sharding is not necessary.
Nginx documentation : http://nginx.org/en/docs/http/ngx_http_spdy_module.html

**Comment by asciimoo** [comment]
@Cqoicebordel i agree, image resizing would be a useful feature, i just don't know how could it be securely implemented

@pointhi yap, good idea, a token could prevent abuse. Requires user session, but it can be handled nonpersistently .

@dalf technically whole searx is an open proxy - a specific one =)

**Comment by dalf** [comment]
The initial issue was fetching image and minimized them. 
There current implementation only fetch the image.

Should we consider the issue closed ?

**Comment by Cqoicebordel** [comment]
I don't think so : the proxy part of the issue was closed using #281.
But let's rename this thread to reflect that it's a discussion about resizing.

**Comment by unixfox** [comment]
I don't think it's a job for Searx to do that. A lot of reverse proxies can cache images and compress them on the fly.
